examples/model_compress/pruning/transformers/transformer_pruning.py

```diff
@@ -0,0 +1,411 @@
+        }
+        ]
+
+    # example 3: prune different layers with different sparsity
+    elif args.usage == 3:
+        attention_name_groups = list(zip(["encoder.layer.{}.attention.self.query".format(i) for i in range(12)],
+                                         ["encoder.layer.{}.attention.self.key".format(i) for i in range(12)],
+                                         ["encoder.layer.{}.attention.self.value".format(i) for i in range(12)],
+                                         ["encoder.layer.{}.attention.output.dense".format(i) for i in range(12)]))
+
+        kwargs = {"ranking_criterion": args.ranking_criterion,
+                  "global_sort": args.global_sort,
+                  "num_iterations": args.num_iterations,
+                  "epochs_per_iteration": args.epochs_per_iteration,
+                  "attention_name_groups": attention_name_groups,
+                  "head_hidden_dim": 64,
+                  "trainer": trainer,
+                  "optimizer": optimizer}
+
+        config_list = [{
+            "sparsity": args.sparsity,
+            "op_types": ["Linear"],
+            "op_names": [x for layer in attention_name_groups[:6] for x in layer]
+        },
+            {
+                "sparsity": args.sparsity / 2,
+                "op_types": ["Linear"],
+                "op_names": [x for layer in attention_name_groups[:6] for x in layer]
+            }
+        ]
+
+    else:
+        raise RuntimeError("Wrong usage number")
+
+    pruner = TransformerHeadPruner(model, config_list, **kwargs)
+    pruner.compress()
+
+    #########################################################################
+    # uncomment the following part to export the pruned model masks
+    # model_path = os.path.join(args.output_dir, "pruned_{}_{}.pth".format(args.model_name, args.task_name))
+    # mask_path = os.path.join(args.output_dir, "mask_{}_{}.pth".format(args.model_name, args.task_name))
+    # pruner.export_model(model_path=model_path, mask_path=mask_path)
+
+    #########################################################################
+    # Speedup
+    # Currently, speeding up Transformers through NNI ModelSpeedup is not supported because of shape inference issues.
+    # However, if you are using the transformers library, you can use the following workaround:
+    # The following code gets the head pruning decisions from the pruner and calls the _prune_heads() function
+    # implemented in models from the transformers library to speed up the model.
+    if args.speed_up:
+        speedup_rules = {}
+        for group_idx, group in enumerate(pruner.attention_name_groups):
+            # get the layer index
+            layer_idx = None
+            for part in group[0].split("."):
+                try:
+                    layer_idx = int(part)
+                    break
+                except:
+                    continue
+            if layer_idx is not None:
```

Comment:
Is there a better way to get the index of the attention head? the first integer may be not strong.

Reply:
This `layer_idx` is the layer index of the BERT encoder. Here I include these lines of code only to show the user how they may take advantage of the `pruned_heads` dict inside pruner to get the pruned heads for each group, and then match each group to the original layer, and finally call the built-in transformers `_prune_heads()` function to do model speedup. This is meant to be a temporary workaround before we can properly handle speedup for transformers.

Reply:
If our speedup code after refactor can handle transformer, then I will replace these lines with our speedup methods (maybe in a separate pr)

Reply:
Since the users are aware of the naming of their own model to prune, I think they can also use their own rules to match layers to groups